GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+        private int bufferContentLength;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        public string ReadMessage()
+        {
+            List<byte> bytes = new List<byte>();
+
+            int bytesRead = this.Read();
+            if (bytesRead == 0)
+            {
+                return null;

```

Comment:
Is this the correct behavior? Or should we assert/throw like we do in the loop below? Either way, it seems like they should be consistent with each other.

Reply:
This one here is correct / expected behavior. This indicates the stream has been closed.

We don't expect the stream to close in the middle of reading a message (e.g. before reading the terminating byte). I will add a code comment here, and tweak the logic in the other location to throw an exception.